### PR TITLE
[RL] Add fp32 lm_head (CastedLinear) for RL training

### PR DESCRIPTION
**Pull request by tianyu-l**
@tianyu-l 's note:
- bitwise between trainer and generator is not affected, as both use the same casting
- e2e loss changes, because lm_head used to run with bf16 now run with fp32 precision.

## Why

RL logprob / KL / advantage math is sensitive to the precision of the lm_head logits. Under bf16 mixed precision the `hidden @ Wᵀ` matmul accumulates over the hidden dim in bf16, which introduces error in the logits — even though the downstream cross-entropy in `compute_logprobs` already upcasts to fp32 (`logits.float()`). The thing that actually needs higher precision is the **matmul accumulation**, not just the softmax.

This is wanted for RL/post-training, not pretraining, so it's opt-in and lives under `experiments/rl`.

## What

- **`CastedLinear`** (`torchtitan/experiments/rl/models/casted_linear.py`): a `Linear` subclass whose forward casts input/weight/bias to a configurable `compute_dtype` (default `float32`) before `F.linear`, returning logits in that dtype. The cast lives in `forward`, so:
  - it's independent of the stored / FSDP-all-gathered parameter dtype, and
  - it's **safe under weight tying** (qwen3 0.6B/1.7B tie embeddings) — the shared parameter keeps its dtype; only the lm_head matmul sees the cast.
- **`LMHeadCastConverter`**: a `ModelConfigConverter` that mirrors `Float8LinearConverter` — walks the model config tree and swaps **only** the `lm_head` node to `CastedLinear.Config`. Every other `Linear` stays plain, so LoRA / quantization converters are unaffected.
- **Wiring**: a `_rl_model_registry` helper in `config_registry.py` appends the converter for **all** RL configs.

## Trainer ↔ generator parity

Because `registry_to_vllm` backs the vLLM generator with the **same** `model_spec`, the generator computes logits via the same `CastedLinear` lm_head (`vllm_wrapper.py: compute_logits`). So the cast applies symmetrically on both sides: trainer (`fp32 FSDP params → bf16 all-gather → fp32 lm_head cast`) and inference (`fp32 trainer → bf16 weight-sync → fp32 lm_head cast`) share the same lossy bf16 round-trip. This **preserves** trainer↔generator bitwise parity for the `*_batch_invariant` configs rather than breaking it. Documented in the converter docstring, with a TODO to investigate removing the lossy `fp32→bf16→fp32` cast pair later.

## Testing

- New CPU unit tests: `torchtitan/experiments/rl/tests/test_casted_linear.py` (5 tests, passing) covering: only-lm_head swap, field preservation, fp32 forward (bitwise-equal to an explicit fp32 matmul and bitwise-*different* from a bf16 matmul), `compute_dtype` configurability, and weight-tying dtype safety.
- `pre-commit run` passes on all changed files.
- Numerics note: this is a *computation* change (bf16→fp32 accumulation), so it is intentionally not bitwise-identical to before. Recommended GPU validation: `test_bitwise_parity.py` on the batch-invariant configs (both sides fp32) + GRPO loss convergence on the off-policy configs.

### CODE REVIEW

File: torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py [hunk 1]
```diff
@@ -31,16 +31,34 @@
...
 _BATCH_INVARIANT_DEBUG = DebugConfig(batch_invariant=True, deterministic=True)
 
 
+def _rl_model_registry(
+    flavor: str,
+    *,
+    attn_backend: str,
+    converters: list[ModelConfigConverter.Config] | None = None,
+) -> ModelSpec:
+    """``qwen3.model_registry`` for RL, with the lm_head fp32 cast always on.
```
Comment (felipemello1) [inline torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py:57]: this doesnt seen to be qwen-specific

File: torchtitan/experiments/rl/models/casted_linear.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        """Forward matmul dtype for the lm_head (key into ``TORCH_DTYPE_MAP``)."""
+
+    def __init__(self, config: Config):
+        self.config = config
+
+    def convert(self, model_config) -> None:
+        for fqn, linear_config, parent, attr in model_config.traverse(Linear.Config):
+            if fqn.rsplit(".", 1)[-1] != self._TARGET:
+                continue
+            new_config = CastedLinear.Config(
```
Comment (shuhuayu) [inline torchtitan/experiments/rl/models/casted_linear.py]: nit: 
```
from dataclasses import fields
new_config = CastedLinear.Config(
      **{f.name: getattr(linear_config, f.name) for f in fields(linear_config)},
      compute_dtype=self.config.compute_dtype,
)
```

File: torchtitan/experiments/rl/models/casted_linear.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""Linear that runs its matmul in a fixed compute dtype, plus the converter
+that swaps the decoder lm_head to it.
+
+RL training (logprob / KL / advantage math) is sensitive to the precision of
+the lm_head logits. With bf16 mixed-precision training the ``hidden @ Wᵀ``
```
Comment (shuhuayu) [inline torchtitan/experiments/rl/models/casted_linear.py]: the accumulation is already in float 32 in bf16 gemm on gpus, the error is in the final step of rounding. maybe say to avoid per-logit bf16 rounding.

File: torchtitan/experiments/rl/models/casted_linear.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+import torch
+import torch.nn.functional as F
+
+from torchtitan.config import TORCH_DTYPE_MAP
+from torchtitan.models.common.nn_modules import Linear
+from torchtitan.protocols.model import ModelConfigConverter
+
+
+class CastedLinear(Linear):
```
Comment (shuhuayu) [inline torchtitan/experiments/rl/models/casted_linear.py]: nit: the past tense of cast is cast itself, so maybe `UpcastLinear` is more direct here?

**Review by shuhuayu — APPROVED** [review]
LGTM!

File: torchtitan/experiments/rl/models/cast_linear.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    goes fp32 (trainer) -> bf16 (weight-sync) -> fp32 (lm_head cast); the
+    trainer's own lm_head input follows the analogous fp32 (FSDP-sharded
+    params) -> bf16 (all-gather) -> fp32 (lm_head cast). Both paths share the
+    same lossy bf16 round-trip, which tends to make trainer<->inference
+    bitwise agreement *easier*, not harder. TODO: investigate whether this
+    fp32->bf16->fp32 cast pair can be removed (keep the weight in fp32 end to
+    end) once that path is supported on both sides.
+    """
+
+    _TARGET = "lm_head"
```
Comment (wwwjn) [inline torchtitan/experiments/rl/models/cast_linear.py:82]: So it will only be single target, not arbitrary target? 

This sounds like to be the MoE router gate can also be put into this converter (not necessarily done in this PR)

File: torchtitan/experiments/rl/models/cast_linear.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    Walks the model config tree and replaces the ``lm_head`` node in place.
+    Targets only the lm_head, so every other Linear stays a plain ``Linear``
+    and LoRA / quantization converters are unaffected.
+
+    Note on trainer/inference bitwise parity: because the same ``model_spec``
+    backs both the trainer and the vLLM generator, the lm_head sees a matched
+    cast chain on both sides. The inference weight, synced from the trainer,
+    goes fp32 (trainer) -> bf16 (weight-sync) -> fp32 (lm_head cast); the
+    trainer's own lm_head input follows the analogous fp32 (FSDP-sharded
+    params) -> bf16 (all-gather) -> fp32 (lm_head cast). Both paths share the
```
Comment (wwwjn) [inline torchtitan/experiments/rl/models/cast_linear.py:75]: The bitwise parity sounds like a coincidence: Both side went throught lossy convertion, but from different reason. For FSDP we can set mixed_precision parameters to control the all-gather dtype, right? Can we also control the weight-sync dtype?

And is bf16 because of weight sync, or because of we set model weights in inference to be bf16 https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/actors/generator.py#L312

Update: Actually I could follow up on this (a end-to-end dtype story in RL), I'm going to try using FSDP for mixed-precision in Generator instead of converting generator's weights all be bf16

File: torchtitan/experiments/rl/models/cast_linear.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    def __init__(self, config: Config):
+        super().__init__(config)
+        self.compute_dtype = TORCH_DTYPE_MAP[config.compute_dtype]
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        bias = None if self.bias is None else self.bias.to(self.compute_dtype)
+        return F.linear(
+            input.to(self.compute_dtype), self.weight.to(self.compute_dtype), bias
+        )
```
Comment (felipemello1) [inline torchtitan/experiments/rl/models/cast_linear.py:60]: noob question: we would have to pay the cast cost on every call, is that right? Can we pay the price only once instead? I don't know if this is expensive or would fragment memory in any way.

File: torchtitan/experiments/rl/models/cast_linear.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+class LMHeadCastConverter(ModelConfigConverter):
+    """Swap the decoder lm_head's ``Linear.Config`` to ``CastedLinear.Config``.
+
+    Walks the model config tree and replaces the ``lm_head`` node in place.
+    Targets only the lm_head, so every other Linear stays a plain ``Linear``
+    and LoRA / quantization converters are unaffected.
+
+    Note on trainer/inference bitwise parity: because the same ``model_spec``
+    backs both the trainer and the vLLM generator, the lm_head sees a matched
```
Comment (felipemello1) [inline torchtitan/experiments/rl/models/cast_linear.py:71]: i haven't checked how we define the model in vLLM, but they also have `head_type` arg: https://docs.vllm.ai/en/stable/api/vllm/config/model/#vllm.config.model.ModelConfig.head_dtype

File: torchtitan/experiments/rl/models/cast_linear.py (hunk 5 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/models/cast_linear.py:82]: i also dont know if some model definition would name it something else, e.g. "head" or "lm_projection". Can we raise an error if TARGET is not found?

**Review by felipemello1 — APPROVED** [review]
thanks! excited to try this out.

I left a couple of questions. On top of them:
1) For my own understanding: after the head is computed in fp32, will all operations after that be in fp32, including softmax?
2) Have you had a chance to produce something like the minimax graph with before and after?

Image

File: torchtitan/experiments/rl/models/cast_linear.py (hunk 7 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/models/cast_linear.py:60]: I'm assuming you mean inference, because in training the weight would be updated by optimizer.

I think that makes sense -- will put a TODO for later investigation.

File: torchtitan/experiments/rl/models/cast_linear.py (hunk 8 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/models/cast_linear.py:71]: It's defined in their `ModelConfig` https://github.com/vllm-project/vllm/blob/main/vllm/config/model.py#L101

I feel it only automatically takes effect in vllm native models.

File: torchtitan/experiments/rl/models/cast_linear.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/models/cast_linear.py:75]: > Can we also control the weight-sync dtype?

From the API calling today, it seems we may not `? https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/actors/generator.py#L894

> And is bf16 because of weight sync, or because of we set model weights in inference to be bf16

These are the same reason, I think.
Inference model weights being in bf16 would cause weight sync to do conversion of state dict from fp32 to bf16. The logic should be in torchstore.
cc @meetv18

File: torchtitan/experiments/rl/models/cast_linear.py (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/models/cast_linear.py:82]: @wwwjn 

> This sounds like to be the MoE router gate can also be put into this converter (not necessarily done in this PR)

You mean https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/moe.py#L264?
I think the idea of using a converter is because it's sometimes used (RL), sometimes not used (pretraining).

For MoE router, it should always be used, so a converter would look strange. I think we might use CastLinear class for MoE router gate though!

Comment (tianyu-l) [reply torchtitan/experiments/rl/models/cast_linear.py:82]: @felipemello1 

> i also dont know if some model definition would name it something else, e.g. "head" or "lm_projection". Can we raise an error if TARGET is not found?

We are running torchtitan model, and `lm_head` is the only name we use in the repo (decoder.py). Good idea of raising if not found.

File: torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py:57]: The `model_registry` call below is from qwen, so claude does it that way.

I don't care too much about this thin wrapper in recipe config space. Different recipes can do whatever they want. We can revisit later.

**Comment by tianyu-l** [comment]
@felipemello1 

> For my own understanding: after the head is computed in fp32, will all operations after that be in fp32, including softmax?

The output will be in fp32. What happens after is up to what we do with it. Concretely, we should do all the loss in fp32. Today this is true
- https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/actors/trainer.py#L48
- https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/trainer.py#L49

> Have you had a chance to produce something like the minimax graph with before and after?

I haven't. It'd be a good followup to do.

File: torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py (hunk 1 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py:57]: should we remove the "qwen.model_registry" from the description though? e.g. if someone uses llama, wouldnt they use this same wrapper?

Comment (tianyu-l) [reply torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py:57]: They wouldn't because qwen's `model_registry` is hardcoded. I can rename the outer function to `qwen_rl_model_registry` or send `model_registry` to this outer function.
